stacktrace.go

```diff
@@ -20,23 +20,72 @@
 
 package zap
 
-import "runtime"
+import (
+	"runtime"
+	"strings"
+	"sync"
+
+	"go.uber.org/zap/internal/bufferpool"
+)
+
+var (
+	defaultUintptrSliceSize = 128
+	uintptrPool             = sync.Pool{
+		New: func() interface{} {
+			return make([]uintptr, defaultUintptrSliceSize)
+		},
+	}
+)
 
 // takeStacktrace attempts to use the provided byte slice to take a stacktrace.
 // If the provided slice isn't large enough, takeStacktrace will allocate
 // successively larger slices until it can capture the whole stack.
-func takeStacktrace(buf []byte, includeAllGoroutines bool) string {
-	if len(buf) == 0 {
-		// We may have been passed a nil byte slice.
-		buf = make([]byte, 1024)
+func takeStacktrace() string {
+	buffer := bufferpool.Get()
```

Comment:
+1 for deferring here, which you've done. Small nit: can we group the defers and the `Get`s, so that acquiring and releasing the resource are on adjacent lines?